File: src/CommandBuffer.cpp
#include "./CommandBuffer.h"

Diligent::RESOURCE_STATE_TRANSITION_MODE rengine_cmdbuffer_get_state(u8 isDeferred) {
	return isDeferred ? Diligent::RESOURCE_STATE_TRANSITION_MODE_VERIFY : Diligent::RESOURCE_STATE_TRANSITION_MODE_TRANSITION;
}

RENGINE void rengine_cmdbuffer_cleardepth(
	Diligent::IDeviceContext* context,
	Diligent::ITextureView* depthStencil,
	Diligent::CLEAR_DEPTH_STENCIL_FLAGS clearFlags,
	float depth,
	u8 stencil,
	u8 isDeferred)
{
	if (!context)
		return;
	context->ClearDepthStencil(
		depthStencil,
		clearFlags,
		depth,
		stencil,
		rengine_cmdbuffer_get_state(isDeferred)
	);
}

RENGINE void rengine_cmdbuffer_clearrt(
	Diligent::IDeviceContext* context,
	Diligent::ITextureView* rt,
	float* color,
	u8 isDeferred
)
{
	if (!context)
		return;

	context->ClearRenderTarget(
		rt,
		color,
	rengine_cmdbuffer_get_state(isDeferred)
	);
}

RENGINE void rengine_cmdbuffer_commitbindings(
	Diligent::IDeviceContext* context,
	Diligent::IShaderResourceBinding* shaderRes,
	u8 isDeferred
)
{
	if (!context)
		return;

	context->CommitShaderResources(shaderRes, rengine_cmdbuffer_get_state(isDeferred));
}

RENGINE void rengine_cmdbuffer_copy_tex(
	Diligent::IDeviceContext* context,
	REngine::CopyTextureInfoDTO* copyInfo,
	u8 isDeferred
)
{
	if (!context)
		return;

	Diligent::CopyTextureAttribs texAttribs;
	texAttribs.pSrcTexture = copyInfo->srcTexture;
	texAttribs.SrcMipLevel = copyInfo->srcMipLevel;
	texAttribs.SrcSlice = copyInfo->srcSlice;
	texAttribs.pSrcBox = copyInfo->srcBox;
	texAttribs.pDstTexture = copyInfo->dstTexture;
	texAttribs.DstMipLevel = copyInfo->dstMipLevel;
	texAttribs.DstSlice = copyInfo->dstSlice;
	texAttribs.DstX = copyInfo->dstX;
	texAttribs.DstY = copyInfo->dstY;
	texAttribs.DstZ = copyInfo->dstZ;
	texAttribs.DstTextureTransitionMode = texAttribs.SrcTextureTransitionMode
		= rengine_cmdbuffer_get_state(isDeferred);

	context->CopyTexture(texAttribs);
}

#if defined(__EMSCRIPTEN__)
RENGINE void rengine_cmdbuffer_draw(
	Diligent::IDeviceContext* context,
	REngine::WebDrawArgs* drawArgs
)
{
	if (!context)
		return;
	Diligent::DrawAttribs attribs;
	attribs.Flags = Diligent::DRAW_FLAG_VERIFY_ALL;
	attribs.NumVertices = drawArgs->numVertices;
	attribs.NumInstances = drawArgs->numInstances;
	attribs.StartVertexLocation = drawArgs->startVertexLocation;
	attribs.FirstInstanceLocation = drawArgs->firstInstanceLocation;
	context->Draw(attribs);
}
#else
RENGINE void rengine_cmdbuffer_draw(
	Diligent::IDeviceContext* context,
	Diligent::DrawAttribs* drawAttribs
)
{
	if (!context)
		return;
	context->Draw(*drawAttribs);
}
#endif

RENGINE void rengine_cmdbuffer_drawindexed(
	Diligent::IDeviceContext* context,
	Diligent::DrawIndexedAttribs* drawAttribs
)
{
	if (!context)
		return;
	context->DrawIndexed(*drawAttribs);
}

RENGINE void* rengine_cmdbuffer_map(
	Diligent::IDeviceContext* context,
	Diligent::IBuffer* buffer,
	Diligent::MAP_TYPE mapType,
	Diligent::MAP_FLAGS mapFlags
)
{
	void* mappedData = null;
	if (!context)
		return mappedData;

	context->MapBuffer(
		buffer,
		mapType,
		mapFlags,
		mappedData
	);
	return mappedData;
}

RENGINE void rengine_cmdbuffer_unmap(
	Diligent::IDeviceContext* context,
	Diligent::IBuffer* buffer,
	Diligent::MAP_TYPE mapType
)
{
	if (!context)
		return;

	context->UnmapBuffer(buffer, mapType);
}

RENGINE void rengine_cmdbuffer_setibuffer(
	Diligent::IDeviceContext* context,
	Diligent::IBuffer* buffer,
	u64 byteOffset,
	u8 isDeferred
)
{
	if (!context)
		return;
	context->SetIndexBuffer(buffer, byteOffset, rengine_cmdbuffer_get_state(isDeferred));
}

RENGINE void rengine_cmdbuffer_setvbuffer(
	Diligent::IDeviceContext* context,
	u32 startSlot,
	u32 numBuffers,
	Diligent::IBuffer** buffers,
	u64* offsets,
	u8 reset,
	u8 isDeferred
)
{
	if (!context)
		return;

	context->SetVertexBuffers(
		startSlot,
		numBuffers,
		buffers,
		offsets,
		rengine_cmdbuffer_get_state(isDeferred),
		reset ? Diligent::SET_VERTEX_BUFFERS_FLAG_RESET : Diligent::SET_VERTEX_BUFFERS_FLAG_NONE
	);
}

RENGINE void rengine_cmdbuffer_setpipeline(
	Diligent::IDeviceContext* context,
	Diligent::IPipelineState* pipeline
)
{
	if (!context)
		return;

	context->SetPipelineState(pipeline);
}

RENGINE void rengine_cmdbuffer_setrts(
	Diligent::IDeviceContext* context,
	Diligent::ITextureView** rts,
	u8 numRts,
	Diligent::ITextureView* depth,
	u8 isDeferred
)
{
	if (!context)
		return;
	context->SetRenderTargets(
		numRts,
		rts,
		depth,
		rengine_cmdbuffer_get_state(isDeferred)
	);
}

RENGINE void rengine_cmdbuffer_updtbuffer(
	Diligent::IDeviceContext* context,
	Diligent::IBuffer* buffer,
	u64 offset,
	u64 size,
	void* ptr,
	u8 isDeferred
) 
{
	if (!context)
		return;
	context->UpdateBuffer(
		buffer,
		offset,
		size,
		ptr,
		rengine_cmdbuffer_get_state(isDeferred)
	);
}

RENGINE void rengine_cmdbuffer_setblendfactors(Diligent::IDeviceContext* context, float r, float g, float b, float a)
{
	if (!context)
		return;

	float colors[] = { r, g, b, a };
	context->SetBlendFactors(colors);
}

RENGINE void rengine_cmdbuffer_setviewports(Diligent::IDeviceContext* context, Diligent::Viewport* viewports, u8 numViewports, u32 rtWidth, u32 rtHeight)
{
	if (!context)
		return;

	context->SetViewports(
		numViewports,
		viewports,
		rtWidth,
		rtHeight
	);
}

RENGINE void rengine_cmdbuffer_setscissors(Diligent::IDeviceContext* context, Diligent::Rect* scissors, u8 numScissors, u32 rtWidth, u32 rtHeight)
{
	if (!context)
		return;

	context->SetScissorRects(
		numScissors,
		scissors,
		rtWidth,
		rtHeight
	);
}

RENGINE void rengine_cmdbuffer_compute(Diligent::IDeviceContext* context, Diligent::DispatchComputeAttribs* attribs)
{
	if (!context)
		return;

	context->DispatchCompute(*attribs);
}

RENGINE void rengine_cmdbuffer_begin_dbg_grp(Diligent::IDeviceContext* context, string* name, float* color)
{
	if (!context)
		return;
	context->BeginDebugGroup(name, color);
}

RENGINE void rengine_cmdbuffer_end_dbg_grp(Diligent::IDeviceContext* context)
{
	if (!context)
		return;
	context->EndDebugGroup();
}

RENGINE void rengine_cmdbuffer_insert_dbg_label(Diligent::IDeviceContext* context, string* label, float* color)
{
	if (!context)
		return;
	context->InsertDebugLabel(label, color);
}

RENGINE void rengine_cmdbuffer_begin(
	Diligent::IDeviceContext* context,
	u32 immediateCtxId
)
{
	if (!context)
		return;
	context->Begin(immediateCtxId);
}

RENGINE void rengine_cmdbuffer_finish_frame(
	Diligent::IDeviceContext* context
)
{
	if (context)
		context->FinishFrame();
}

RENGINE Diligent::ICommandList* rengine_cmdbuffer_finish_command_list(
	Diligent::IDeviceContext* context
)
{
	if (!context)
		return null;
	Diligent::ICommandList* commandList;
	context->FinishCommandList(&commandList);
	return commandList;
}

RENGINE void rengine_cmdbuffer_transition_shader_resources(
	Diligent::IDeviceContext* context,
	Diligent::IPipelineState* pipelineState,
	Diligent::IShaderResourceBinding* shaderResourceBinding
)
{
	if (!context)
		return;
	context->TransitionShaderResources(
		pipelineState,
		shaderResourceBinding
	);
}

RENGINE void rengine_cmdbuffer_set_stencil_ref(
	Diligent::IDeviceContext* context,
	u32 stencilRef
)
{
	if (!context)
		return;
	context->SetStencilRef(stencilRef);
}

RENGINE void rengine_cmdbuffer_invalidate_state(
	Diligent::IDeviceContext* context
)
{
	if (context)
		context->InvalidateState();
}

RENGINE void rengine_cmdbuffer_next_subpass(
	Diligent::IDeviceContext* context
)
{
	if (context)
		context->NextSubpass();
}

RENGINE void rengine_cmdbuffer_generate_mips(
	Diligent::IDeviceContext* context,
	Diligent::ITextureView* textureView
)
{
	if (context)
		context->GenerateMips(textureView);
}

RENGINE void rengine_cmdbuffer_transition_resource_states(
	Diligent::IDeviceContext* context,
	u32 barrierCount,
	Diligent::StateTransitionDesc* resourceBarriers
)
{
	if (context)
		context->TransitionResourceStates(barrierCount, resourceBarriers);
}

RENGINE void rengine_cmdbuffer_resolve_texture_subresource(
	Diligent::IDeviceContext* context,
	Diligent::ITexture* srcTexture,
	Diligent::ITexture* dstTexture,
	Diligent::ResolveTextureSubresourceAttribs* resolveAttribs
)
{
	if (context)
		context->ResolveTextureSubresource(srcTexture, dstTexture, *resolveAttribs);
}

RENGINE void rengine_cmdbuffer_exec_command_list(
	Diligent::IDeviceContext* context,
	u32 numCommandLists,
	Diligent::ICommandList** list
)
{
	if (!context || numCommandLists ==0 || list == null)
		return;
	context->ExecuteCommandLists(numCommandLists, list);
}

File: jslib/command-buffer.js
const { getPtrSize, malloc, free, compileWriteInstructions } = require("./core");
const { getModule } = require("./native-module");
const { NativeObject } = require("./native-object");
const { PipelineState } = require("./pipeline-state");
const { TextureView } = require("./texture-view");

/**
 * 
 * @param {number} colorPtr 
 * @param {number[]} colors 
 */
function _fillColor(colorPtr, colors) {
    const ptrSize = getPtrSize();
    const heap = getModule().HEAPF32;
    for(let i =0; i < 4; ++i) {
        const ptr = colorPtr + (i * ptrSize);
        heap[(ptr >> 2)] = i < colors.length ? colors[i] : 0; 
    }
}
/**
 * 
 * @param {number} rtPtr 
 * @param {TextureView[]} rts 
 */
function _setRt(rtPtr, rts) {
    const ptrSize = getPtrSize();
    const heap = getModule().HEAPU32;
    rts.forEach(x => {
        heap[(rtPtr >> 2)] = x.handle;
        rtPtr += ptrSize;
    });
}

const $drawArgsPayload = Symbol('__payload');
class DrawArgs {
    constructor() {
        const payload = {
            numVertices : 0,
            numInstances : 1,
            startVertexLocation: 0,
            firstInstanceLocation: 0,
            dirty : true,
            disposed : false,
            ptr : malloc(4 * getPtrSize()),
            writeCall : ()=> void(0)
        };

        const instructions = [
            [()=> payload.numVertices, 'uint'],
            [()=> payload.numInstances, 'uint'],
            [()=> payload.startVertexLocation, 'uint'],
            [()=> payload.firstInstanceLocation, 'uint'],
        ];
        payload.writeCall = compileWriteInstructions(instructions, payload.ptr);
 
        this[$drawArgsPayload] = payload;
    }

    dispose() {
        const payload = this[$drawArgsPayload];
        if(payload.disposed)
            return;
        payload.disposed = true;
        free(payload.ptr);
    }

    toPtr() {
        const payload = this[$drawArgsPayload];
        if(payload.disposed)
            return 0;
        if(payload.dirty)
        {
            payload.dirty = false;
            payload.writeCall();
        }
        return payload.ptr;
    }

    get numVertices() { return this[$drawArgsPayload].numVertices; }
    set numVertices(value) {
        const payload = this[$drawArgsPayload];
        payload.numVertices = value;
        payload.dirty = true;
    }

    get numInstances() { return this[$drawArgsPayload].numInstances; }
    set numInstances(value) {
        const payload = this[$drawArgsPayload];
        payload.numInstances = value;
        payload.dirty = true;
    }

    get startVertexLocation() { return this[$drawArgsPayload].startVertexLocation; }
    set startVertexLocation(value) {
        const payload = this[$drawArgsPayload];
        payload.startVertexLocation = value;
        payload.dirty = true;
    }

    get firstInstanceLocation() { return this[$drawArgsPayload].firstInstanceLocation; }
    set firstInstanceLocation(value) {
        const payload = this[$drawArgsPayload];
        payload.firstInstanceLocation = value;
        payload.dirty = true;
    }
}
class CommandBuffer {
    /**
     * Create command buffer from command buffer ptr
     * @param {number} handle 
     */
    constructor(handle) {
        this.disposed = false;
        this.handle = handle;
        this._colorCopyBuffer = getModule()._malloc(4 * getPtrSize());
        this._rtCopyBuffer = getModule()._malloc(16 * getPtrSize());
    }

    dispose() {
        if(this.disposed)
            return;
        this.disposed = true;
        getModule()._free(this._colorCopyBuffer);
        getModule()._free(this._rtCopyBuffer);
        getModule()._rengine_object_releaseref(this.handle);
        
        this._colorCopyBuffer = this._rtCopyBuffer = 0;
    }
    /**
     * 
     * @param {TextureView} depthStencil 
     * @param {number} clearFlags 
     * @param {number} depth 
     * @param {number} stencil 
     * @returns {CommandBuffer} self instance
     */
    clearDepth(depthStencil, clearFlags, depth, stencil) {
        if(this.disposed || depthStencil.handle == 0)
            return this;
        
        getModule()._rengine_cmdbuffer_cleardepth(
            this.handle,
            depthStencil.handle,
            clearFlags,
            depth,
            stencil
        );

        return this;
    }
    /**
     * 
     * @param {TextureView} rt 
     * @param {number[]} colorArr 
     * @returns {CommandBuffer} self instance
     */
    clearRt(rt, colorArr) {
        if(this.disposed || rt.handle == 0)
            return this;
        _fillColor(this._colorCopyBuffer, colorArr);
        getModule()._rengine_cmdbuffer_clearrt(
            this.handle,
            rt.handle,
            this._colorCopyBuffer,
            0x0
        );

        return this;
    }

    commitBindings(srb) {
        throw new Error('not implemented');
    }
    copy(copyInfo) {
        throw new Error('not implemented');
    }
    /**
     * 
     * @param {DrawArgs} drawArgs 
     */
    draw(drawArgs) {
        if(this.disposed || !drawArgs)
            return this;
        getModule()._rengine_cmdbuffer_draw(this.handle, drawArgs.toPtr());
        return this;
    }
    drawIndexed(drawArgs) {
        throw new Error('not implemented');
    }
    map(buffer, mapType, mapFlags) {
        throw new Error('not implemented');
    }
    unmap(buffer, mapType) {
        throw new Error('not implemented');
    }
    updateBuffer(buffer, offset, data) {
        throw new Error('not implemented');
    }
    /**
     * 
     * @param {*} buffer 
     * @param {bool} reset 
     */
    setVertexBuffer(buffer, reset) {
        throw new Error('not implemented');
    }
    setVertexBuffers(startSlot, buffers, offsets, reset) {
        throw new Error('not implemented');
    }
    setIndexBuffer(buffer, byteOffset) {
        throw new Error('not implemented');
    }
    /**
     * 
     * @param {PipelineState} pipelineState 
     */
    setPipeline(pipelineState) {
        if(this.disposed || pipelineState.disposed)
            return this;
        getModule()._rengine_cmdbuffer_setpipeline(this.handle, pipelineState.handle);
        return this;
    }
    /**
     * 
     * @param {TextureView} rt 
     * @param {TextureView?} depthStencil 
     * @returns {CommandBuffer} self instance
     */
    setRt(rt, depthStencil) {
        if(this.disposed || rt.disposed)
            return this;
        _setRt(this._rtCopyBuffer, [rt]);
        getModule()._rengine_cmdbuffer_setrts(
            this.handle,
            this._rtCopyBuffer,
            1,
            depthStencil?.handle ?? 0,
            0x0);
        return this;
    }
    /**
     * 
     * @param {TextureView[]} rts render targets 
     * @param {TextureView?} depthStencil 
     */
    setRts(rts, depthStencil) {
        if(this.disposed || rts.length == 0)
            return this;
        if(rts[0].handle == 0)
            return this;
        _setRt(this._rtCopyBuffer, rts);
        getModule()._rengine_cmdbuffer_setrts(
            this.handle,
            this._rtCopyBuffer,
            rts.length,
            depthStencil?.handle ?? 0,
            0x0
        );
        return this;
    }
    setBlendFactors(r, g, b, a) {
        if(this.disposed)
            return this;
        getModule()._rengine_cmdbuffer_setblendfactors(
            this.handle,
            r / 255.0,
            g / 255.0,
            b / 255.0,
            a / 255.0
        );
        return this;
    } 
    setViewports(viewports, rtWidth, rtHeight) {
        throw new Error('not implemented');
    }
    setViewport(viewport, rtWidth, rtHeight) {
        throw new Error('not implemented');
    }
    setScissors(scissors, rtWidth, rtHeight) {
        throw new Error('not implemented');
    }
    setScissor(scissor, rtWidth, rtHeight) {
        throw new Error('not implemented');
    }
    compute(args) {
        throw new Error('not implemented');
    }
    setStencilRef(stencilRef) {
        if(this.disposed)
            return this;
        getModule()._rengine_cmdbuffer_set_stencil_ref(this.handle, stencilRef);
    }
    invalidateState() {
        return this;
    }
    nextSubpass() {
        return this;
    }
    generateMips(texView) {
        throw new Error('not implemented');
    }
    transitionResourceState(resourceBarriers) {
        throw new Error('not implemented');
    }
    resolveTextureSubresource(srcTexture, dstTexture, resolveDesc) {
        throw new Error('not implemented');
    }
    begin() {
        return this;
    }
    finishFrame() {
        return this;
    }
    finishCommandList() {
        return this;
    }
    executeCommandList(list) {
        return this;
    }
}

class NullCommandBuffer {
    constructor(){
        this.handle = 0;
        this.disposed = false;
        const instance = this;
        [
            'clearDepth',
            'clearRt',
            'commitBindings',
            'copy',
            'draw',
            'drawIndexed',
            'unmap',
            'updateBuffer',
            'setVertexBuffer',
            'setVertexBuffers',
            'setIndexBuffer',
            'setPipeline',
            'setRt',
            'setRts',
            'setBlendFactors',
            'setViewports',
            'setViewport',
            'setScissors',
            'setScissor',
            'compute',
            'setStencilRef',
            'invalidateState',
            'nextSubpass',
            'generateMips',
            'transitionResourceState',
            'resolveTextureSubresource',
            'begin',
            'finishFrame',
            'finishCommandList',
            'executeCommandList'
        ].forEach(x => {
            instance[x] = (...args)=> this;
        });
    }
    dispose() {}
    map(buffer, mapType, mapFlags) {
        return 0;
    }
}

const _nullCommandBuffer = new NullCommandBuffer();
function getNullCommandBuffer() { return _nullCommandBuffer; }

module.exports = {
    DrawArgs,
    CommandBuffer,
    getNullCommandBuffer
};